File: TwoSum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer,Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            map.put(nums[i],i);
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(target - nums[i]) && map.get(target - nums[i]) != i) {
                return new int[] {i,map.get(target - nums[i])};
            }
        }
        return null;
    }
}

//map.containsKey()时间复杂度为O(1),若hashcode设计不合理，则复杂度会为O(n)

File: Array/TwoSum.scala
object Solution {
    def twoSum(nums: Array[Int], target: Int): Array[Int] = {
        val map = new scala.collection.mutable.HashMap[Int,Int];
        for (i <- 0 until nums.length) {
            map(nums(i)) = i;
        }
        for (i <- 0 until nums.length) {
            if (map.contains(target - nums(i)) && map(target - nums(i)) != i) {
                return Array(i,map(target - nums(i)));
            }
        }
        return null;
    }
}

